### PR TITLE
Fix crash in code model where we were blindly casting to an interface.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Paging @DustinCampbell

File: src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs [hunk 1]
```diff
@@ -17,14 +17,17 @@ public NodeKeyValidation()
 
         public void AddProject(AbstractProject project)
         {
-            var provider = (IProjectCodeModelProvider)project;
-            var fcms = provider.ProjectCodeModel.GetCachedFileCodeModelInstances();
-
-            foreach (var fcm in fcms)
+            var provider = project as IProjectCodeModelProvider;
+            if (provider != null)
```
Comment (jaredpar) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: What is the implication if the project doesn't implement `IProjectCodeModelProvider`?  Is it a bug that we should be logging?

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: it means it just doesn't participate in codemodely things.  i.e. you can't get the codemodel view of the files in this project.  TypeScript is such an example.  We simply do not participate in CodeModel at all (and have no intent to for the forseeable future).

**Comment by CyrusNajmabadi** [comment]
Jenkind retest this please.

File: src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs (hunk 1 above)
Comment (DustinCampbell) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: @jaredpar, ```IProjectCodeModelProvider``` is actually just an abstraction that allows us to push our CodeModel stuff into into another assembly. It is wholly ours.

@CyrusNajmabadi, when/why did TypeScript start using our internal project system bits? I didn't know that was a thing. I though TypeScript didn't have a formal project system.

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: This is how we tie into Roslyn's workspace.  It provides the glue that maps between things like Documents and the IVsHierarchys and ItemIds that we *do* use.  Without this there is no correspondence between a random file in Solution explorer and a Roslyn document.  Without this there is no hookup between VS' editor fiters, and Roslyn's command handlers.  etc. etc.

"I though TypeScript didn't have a formal project system."

You are correct in that we don't have an *ms build* project system.  But we do have a project system.  It's just file system based.  i.e. we'll grovel your file system to determine the files that are in your 'virtual project'.  However, once we make this, we do need to both present a UI in solution explorer for this, and we need to hook into roslyn.  Using your roslyn project system bits makes this exceedingly simple.

By and large those project system bits are resilient and make no assumptions that the impl will be C# or VB.  But occasionally there is something like this that leaks through.

Comment (DustinCampbell) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: Ah that makes. I didn't realize you had to use the VS workspace stuff, so I thought is was new. Makes sense!

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs]: Ideally we could possibly do this without needing to directly access these specific classes.  @jasonmalinowski might have ideas on where we can go here in the future.  But this was by far the lowest cost way to be able to integrate with valuable infrastructure pieces like the VisualStudioProjectTracker.    Without that, we'd likely have had to write thousands of lines of glue code.  And that makes :panda_face: :sob:

**Comment by HellBrick** [comment]
Just for the reference: it's supposed to fix https://github.com/Microsoft/TypeScript/issues/3957.

Is there any info on how or when this fix will be delivered to the VS users? And is there any workaround that can be used to repair the rename experience until it's delivered?

**Comment by jasonmalinowski** [comment]
Late to the party but :+1:.

@HellBrick: @CyrusNajmabadi or @paulvanbrenk would best know if there's a workaround available for your issue.

**Comment by manifestarium** [comment]
oh my god, there's no workaround for freakin typescript and name refactoring.  jesus christ, why do i keep putting off moving to linux

**Pull request merged**